### PR TITLE
LinearExecutionTraversal: Add an option to connect adjacent code, use in SimplifyLocals

### PR DESCRIPTION
**Pull request by kripken**
This addresses most of the minor regression from the correctness fix in #5857.
That PR makes us consider calls as branching, but in some cases it is ok to
ignore that branching (see the comment in the code), which this PR allows as
an option.

This undoes one test change from that PR (showing it undoes the regression).
More tests are added to cover this specifically as well.

### CODE REVIEW

File: test/lit/passes/type-ssa_and_merging.wast [hunk 1]
```diff
@@ -22,6 +22,10 @@
   ;; NOP:      (type $none_=>_i32 (func (result i32)))
 
   ;; NOP:      (import "a" "b" (func $import (type $none_=>_i32) (result i32)))
+  ;; YES:       (type $A$2 (sub $A (struct )))
+
+  ;; YES:       (type $A$1 (sub $A (struct )))
+
```
Comment (kripken) [inline test/lit/passes/type-ssa_and_merging.wast:28]: This is unrelated, and is a change I've noticed for a while now but didn't have time to look into. I think perhaps we just updated this file manually and not with the auto updater. This is what the auto updater adds for it.

File: src/ir/linear-execution.h [hunk 2]
```diff
@@ -44,14 +44,49 @@ struct LinearExecutionWalker : public PostWalker<SubType, VisitorType> {
     self->noteNonLinear(*currp);
   }
 
+  // Optionally, we can connect adjacent basic blocks. "Adjacent" here means
```
Comment (aheejin) [inline src/ir/linear-execution.h:47]: This can be basically extended to any dominator-dominatee relationship, but given that this is a linear execution walker, we limit the scope to the adjacent blocks, correct?

I thought about whether "consecutive" sounds better, but usually this word doesn't contain the meaning that the first can branch out or return, so I'm not sure.

File: src/passes/SimplifyLocals.cpp [hunk 3]
```diff
@@ -998,6 +998,11 @@ struct SimplifyLocals
     struct EquivalentOptimizer
       : public LinearExecutionWalker<EquivalentOptimizer> {
 
+      // It is ok to look at adjacent blocks together, as if a later part of a
+      // block is not reached that is fine - changes we make there would not be
+      // reached in that case.
+      bool connectAdjacentBlocks = true;
```
Comment (aheejin) [inline src/passes/SimplifyLocals.cpp:1004]: Are the cases where we need this the same as where we need `bool ignoreBranchesOutsideOfFunc = true;` for `cfg-traversal.h`?

**Review by aheejin — APPROVED** [review]

File: src/ir/linear-execution.h (hunk 2 above)
Comment (kripken) [reply src/ir/linear-execution.h:47]: Yes, exactly, it's the simple case during linear execution when we see dominator-dominatee.

File: src/passes/SimplifyLocals.cpp (hunk 3 above)
Comment (kripken) [reply src/passes/SimplifyLocals.cpp:1004]: I think it's a little different. I was actually thinking of saying in a comment "this is like ignoreBranchesOutsideOfFunc" but it seemed like it could be more confusing than helpful, though I'm not sure...

The main difference is there we ignore branches outside of the function. We never ignore a branch inside it. So we use that in cases where we just care about state in locals, for example, that goes away when we branch out of the function. We can use that for local coalescing for that reason. But we couldn't use this PR for local coalescing since we need to be aware of every branch inside the function, or else we could miss some interferences.

Comment (aheejin) [reply src/passes/SimplifyLocals.cpp:1004]: Isn't this PR also about excluding branches out of a function? Ah, I guess it is branch within a function in case of `if`..?

Comment (kripken) [reply src/passes/SimplifyLocals.cpp:1004]: Yes, exactly, an if would be inside. I think I can also handle `br_if` etc. but I'll leave that for later.
